### PR TITLE
fix web csm, barrier generation, metal buffer/descriptor upload

### PR DESCRIPTION
**Pull request by star-e**
Updated custom pipeline.
fix web csm rendering problem.
fgdispatcher now generates correct barriers.
metal walkaround for buffer/descriptors update between render pass begin/end

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

File: native/cocos/renderer/gfx-metal/MTLDevice.h [hunk 1]
```diff
@@ -86,6 +86,10 @@ class CCMTLDevice final : public Device {
             _swapchains.erase(iter);
         }
     }
+    
+    CommandBuffer* transferCommandBuffer() const;
+    
+    void writeBuffer(Buffer* buffer, const void* data, size_t size);
```
Comment (minggo) [inline native/cocos/renderer/gfx-metal/MTLDevice.h]: Why not use uint32_t?

Comment (star-e) [reply native/cocos/renderer/gfx-metal/MTLDevice.h]: @hana-alice

Comment (hana-alice) [reply native/cocos/renderer/gfx-metal/MTLDevice.h]: webgpu decorate buffer size with uint64_t, so try change buffer size to try keep consistent here . 
if size_t is not needed: https://github.com/star-e/cocos-engine/pull/96

Comment (minggo) [reply native/cocos/renderer/gfx-metal/MTLDevice.h]: I think uint32_t is enough as our engine uses uint32_t for other codes.

**Comment by star-e** [comment]
@cocos-robot run test cases

File: native/cocos/renderer/gfx-metal/MTLDevice.h (hunk 1 above)
Comment (star-e) [reply native/cocos/renderer/gfx-metal/MTLDevice.h]: pr has been merged. Now metal uses uint32_t

**Review by minggo — APPROVED** [review]

**Pull request merged**